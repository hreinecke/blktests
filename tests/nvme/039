#!/bin/bash
# SPDX-License-Identifier: GPL-3.0+
# Copyright (C) 2021 Hannes Reinecke, SUSE Labs
#
# Create authenticated connections

. tests/nvme/rc

DESCRIPTION="Create authenticated connections"
QUICK=1

requires() {
	_nvme_requires
	_have_modules loop
	_require_nvme_trtype_is_fabrics
	_require_nvme_cli_auth
}


test() {
	local port
	local subsys="blktests-subsystem-1"
	local hostid="$(uuidgen)"
	local hostnqn="nqn.2014-08.org.nvmexpress:uuid:${hostid}"
	local scratch="/tmp/blktest-ns1.img"
	local hostkey
	local ctrldev

	echo "Running ${TEST_NAME}"

	hostkey="$(nvme gen-dhchap-key -n ${subsys} 2> /dev/null)"
	if [ $? -ne 0 ] ; then
		echo "nvme gen-dhchap-key command missing"
		return 1
	fi

	_setup_nvmet

	truncate -s 512M "${scratch}"

	port="$(_create_nvmet_port "${nvme_trtype}")"

	_create_nvmet_subsystem "${subsys}" "${scratch}"
	_add_nvmet_subsys_to_port "${port}" "${subsys}"
	_create_nvmet_host "${subsys}" "${hostnqn}" "${hostkey}"

	# Test unauthenticated connection (should fail)
	echo "Test unauthenticated connection"
	_nvme_connect_subsys "${nvme_trtype}" "${subsys}" \
			     "${def_traddr}" "${def_trscvid}" \
			     "${hostnqn}" "${hostid}"

	ctrldev=$(_find_nvme_dev "${subsys}")
	if [ -n "$ctrldev" ] ; then
		echo "nvme controller found"
	fi

	_nvme_disconnect_subsys "${subsys}"

	# Test authenticated connection
	echo "Test authenticated connection"
	_nvme_connect_subsys "${nvme_trtype}" "${subsys}" \
			     "${def_traddr}" "${def_trscvid}" \
			     "${hostnqn}" "${hostid}" "${hostkey}"

	ctrldev=$(_find_nvme_dev "${subsys}")
	if [ -z "$ctrldev" ] ; then
		echo "nvme controller not found"
	fi

	_nvme_disconnect_subsys "${subsys}"

	_remove_nvmet_subsystem_from_port "${port}" "${subsys}"
	_remove_nvmet_subsystem "${subsys}"

	_remove_nvmet_port "${port}"

	_remove_nvmet_host "${hostnqn}"

	rm ${scratch}

	echo "Test complete"
}
