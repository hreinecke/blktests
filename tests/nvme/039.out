Running nvme/039
ANA state failback
1: grpid 1 state optimized
1: grpid 2 state inaccessible
2: grpid 1 state inaccessible
2: grpid 2 state optimized
ANA state change
1: grpid 1 state change
1: grpid 2 state change
2: grpid 1 state change
2: grpid 2 state change
ANA state failover
1: grpid 1 state inaccessible
1: grpid 2 state optimized
2: grpid 1 state optimized
2: grpid 2 state inaccessible
NQN:blktests-subsystem-1 disconnected 2 controller(s)
Test complete
